lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,431 @@
+}
+
+/// Validates rules for field names and value types, coercing in some cases.
+fn coerce_field_names_and_values(
+    mut log: LogEvent,
+    missing_prefix: &mut Vec<String>,
+) -> vector_core::Result<LogEvent> {
+    if let Some(event_data) = log.as_map_mut() {
+        for (field, value) in event_data.iter_mut() {
+            match field.as_str() {
+                VERSION | HOST | SHORT_MESSAGE | FULL_MESSAGE | FACILITY | FILE => {
+                    if !value.is_bytes() {
+                        err_invalid_type(field, "UTF-8 string", value.kind_str())?;
+                    }
+                }
+                TIMESTAMP => {
+                    if !(value.is_timestamp() || value.is_integer()) {
+                        err_invalid_type(field, "timestamp or integer", value.kind_str())?;
+                    }
+                }
+                LEVEL => {
+                    if !value.is_integer() {
+                        err_invalid_type(field, "integer", value.kind_str())?;
+                    }
+                }
+                LINE => {
+                    if !(value.is_float() || value.is_integer()) {
+                        err_invalid_type(field, "number", value.kind_str())?;
+                    }
+                }
+                _ => {
+                    // Additional fields must be prefixed with underscores.
+                    // Prepending the underscore since vector adds fields such as 'source_type'
+                    // which would otherwise throw errors.
+                    if !field.is_empty() && !field.starts_with('_') {
+                        // flag the field as missing prefix to be modified later
+                        missing_prefix.push(field.to_string());
+                    }
+
+                    // additional fields must be only word chars, dashes and periods.
+                    if !VALID_FIELD_REGEX.is_match(field) {
+                        return MissingFieldSnafu { field }
+                            .fail()
+                            .map_err(|e| e.to_string().into());
+                    }
+
+                    // additional field values must be only strings or numbers
+                    if !(value.is_integer() || value.is_float() || value.is_bytes()) {
+                        err_invalid_type(field, "string or number", value.kind_str())?;
+                    }
+                }
+            }
+        }
+    }
+    Ok(log)
+}
+
+/// Validate the input log event is valid GELF, potentially coercing the event into valid GELF
+fn to_gelf_event(log: LogEvent) -> vector_core::Result<LogEvent> {
+    let mut missing_prefix = vec![];
+    let mut log = coerce_required_fields(log)
+        .and_then(|log| coerce_field_names_and_values(log, &mut missing_prefix))?;
+
+    // rename additional fields that were flagged as missing the underscore prefix
+    for field in missing_prefix {
+        log.rename_key(path!(field.as_str()), &*format!("_{}", &field));
+    }
```

Comment:
Not a big deal, but could the missing prefix handling also be `and_then()`d above?